homeassistant/components/sensor/gtfs.py

```diff
@@ -48,73 +50,143 @@
+    now_date = now.strftime(DATE_FORMAT)
+    yesterday = now - datetime.timedelta(days=1)
+    yesterday_date = yesterday.strftime(DATE_FORMAT)
+    tomorrow = now + datetime.timedelta(days=1)
+    tomorrow_date = tomorrow.strftime(DATE_FORMAT)
 
     from sqlalchemy.sql import text
 
-    sql_query = text("""
-    SELECT trip.trip_id, trip.route_id,
-           time(origin_stop_time.arrival_time) AS origin_arrival_time,
-           time(origin_stop_time.departure_time) AS origin_depart_time,
-           origin_stop_time.drop_off_type AS origin_drop_off_type,
-           origin_stop_time.pickup_type AS origin_pickup_type,
-           origin_stop_time.shape_dist_traveled AS origin_shape_dist_traveled,
-           origin_stop_time.stop_headsign AS origin_stop_headsign,
-           origin_stop_time.stop_sequence AS origin_stop_sequence,
-           time(destination_stop_time.arrival_time) AS dest_arrival_time,
-           time(destination_stop_time.departure_time) AS dest_depart_time,
-           destination_stop_time.drop_off_type AS dest_drop_off_type,
-           destination_stop_time.pickup_type AS dest_pickup_type,
-           destination_stop_time.shape_dist_traveled AS dest_dist_traveled,
-           destination_stop_time.stop_headsign AS dest_stop_headsign,
-           destination_stop_time.stop_sequence AS dest_stop_sequence
-    FROM trips trip
-    INNER JOIN calendar calendar
-               ON trip.service_id = calendar.service_id
-    INNER JOIN stop_times origin_stop_time
-               ON trip.trip_id = origin_stop_time.trip_id
-    INNER JOIN stops start_station
-               ON origin_stop_time.stop_id = start_station.stop_id
-    INNER JOIN stop_times destination_stop_time
-               ON trip.trip_id = destination_stop_time.trip_id
-    INNER JOIN stops end_station
-               ON destination_stop_time.stop_id = end_station.stop_id
-    WHERE calendar.{day_name} = 1
-               AND origin_depart_time > time(:now_str)
-    AND start_station.stop_id = :origin_station_id
-               AND end_station.stop_id = :end_station_id
-    AND origin_stop_sequence < dest_stop_sequence
-    AND calendar.start_date <= :today
-    AND calendar.end_date >= :today
-    ORDER BY origin_stop_time.departure_time
-    LIMIT 1
-    """.format(day_name=day_name))
-    result = sched.engine.execute(sql_query, now_str=now_str,
+    # Fetch all departures for yesterday, today and optionally tomorrow,
+    # up to an overkill maximum in case of a departure every minute for those
+    # days.
+    limit = 24 * 60 * 60 * 2
+    tomorrow_select = tomorrow_where = tomorrow_order = ''
+    if include_tomorrow:
+        limit = limit / 2 * 3
+        tomorrow_name = tomorrow.strftime('%A').lower()
+        tomorrow_select = "calendar.{} AS tomorrow,".format(tomorrow_name)
+        tomorrow_where = "OR calendar.{} = 1".format(tomorrow_name)
+        tomorrow_order = "calendar.{} DESC,".format(tomorrow_name)
+
+    sql_query = """
```

Comment:
The updated full query captures all possible data points from yesterday, today and tomorrow. The results are then parsed to create an accurate timetable map for all of today's departure times, and optionally tomorrow's too.